src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingService.cs

```diff
@@ -73,13 +79,143 @@ internal static LanguageService LanguageServiceInstance
+            serviceHost.SetRequestHandler(ScriptingListObjectsRequest.Type, this.HandleListObjectsRequest);
+
+            // Register handler for shutdown event
+            serviceHost.RegisterShutdownTask((shutdownParams, requestContext) =>
+            {
+                this.Dispose();
+                return Task.FromResult(0);
+            });
+        }
+
+        /// <summary>
+        /// Handles request to execute start the list objects operation.
+        /// </summary>
+        private async Task HandleListObjectsRequest(ScriptingListObjectsParams parameters, RequestContext<ScriptingListObjectsResult> requestContext)
+        {
+            try
+            {
+                ScriptingListObjectsOperation operation = new ScriptingListObjectsOperation(parameters, requestContext);
+                RunListObjectsTask(operation);
+                await requestContext.SendResult(new ScriptingListObjectsResult { OperationId = operation.OperationId });
+            }
+            catch (Exception e)
+            {
+                await requestContext.SendError(e);
+            }
+        }
+
+        /// <summary>
+        /// Handles request to execute start the script operation.
+        /// </summary>
+        public async Task HandleScriptExecuteRequest(ScriptingParams parameters, RequestContext<ScriptingResult> requestContext)
+        {
+            try
+            {
+                ScriptingScriptOperation operation = new ScriptingScriptOperation(parameters, requestContext);
+                RunScriptTask(operation);
+                await requestContext.SendResult(new ScriptingResult { OperationId = operation.OperationId });
+            }
+            catch (Exception e)
+            {
+                await requestContext.SendError(e);
+            }
+        }
+
+        /// <summary>
+        /// Handles request to cancel the database script operation
+        /// </summary>
+        public async Task HandleScriptCancelRequest(ScriptingCancelParams parameters, RequestContext<ScriptingCancelResult> requestContext)
+        {
+            try
+            {
+                ScriptingOperation operation = null;
+                if (this.ActiveOperations.TryRemove(parameters.OperationId, out operation))
+                {
+                    operation.Cancel();
+                }
+
+                await requestContext.SendResult(new ScriptingCancelResult());
+            }
+            catch (Exception e)
+            {
+                await requestContext.SendError(e);
+            }
+        }
+
+        /// <summary>
+        /// Runs the async task to list scriptable objects.
+        /// </summary>
+        private void RunListObjectsTask(ScriptingListObjectsOperation operation)
+        {
+            Task.Run(async () =>
+            {
+                try
+                {
+                    this.ActiveOperations[operation.OperationId] = operation;
```

Comment:
A new operation is instantiated per request, and each has a unique operation id so this should never happen. I added debug assert to make this more explicit.